nlmod/dims/layers.py

```diff
@@ -2003,3 +2006,326 @@ def get_isosurface(da, z, value, input_core_dims=None, exclude_dims=None, **kwar
+            return np.nan  # observation below or above model boundaries
+        if mtop == mbot:
+            return mtop  # screen top and bot in same layer
+        # find modellayer with the longest screen length
+        ftop = ds_obs["screen_top"].values[i]
+        fbot = ds_obs["screen_bot"].values[i]
+        botm_single_obs = (
+            ds_obs["botm"]
+            .isel(**{dimname: i}, layer=range(int(mtop - 1), int(mbot + 1)))
+            .values
+        )
+        botm_single_obs[0] = ftop
+        botm_single_obs[-1] = fbot
+        return np.argmin(np.diff(botm_single_obs)) + mtop
+
+    modellayer = [get_max_overlap_model_layer(i) for i in range(ds_obs.sizes[dimname])]
+
+    return modellayer
+
+
+def get_modellayers_indexer(
+    ds,
+    df,
+    x="x",
+    y="y",
+    screen_top="screen_top",
+    screen_bottom="screen_bottom",
+    full_output=False,
+    drop_nan_layers=False,
+):
+    """Get a model layer indexer (dataset) for a dataframe with observation wells.
+
+    The dataframe must contain the spatial data of the observation wells, i.e.
+    the x, y coordinates and the screen top and bottom values. The column names
+    corresponding to these values can be specified with keyword arguments.
+
+    Note that the returned x and y data (if grid is structured) are the cell
+    coordinates and not the original coordinates of the observation points!
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Model Dataset containing layer elevations (top, botm).
+    df : pd.DataFrame
+        DataFrame with x, y coordinates and screen_top and screen_bottom values.
+    x : str, optional
+        name of the x-coordinate column in df. The default is "x".
+    y : str, optional
+        name of the y-coordinate column in df. The default is "y".
+    screen_top : str, optional
+        name of the screen top column in df. The default is "screen_top".
+    screen_bottom : str, optional
+        name of the screen bottom column in df. The default is "screen_bottom".
+    full_output : bool, optional
+        If True, return all variables needed to construct the model layer indexer.
+        If False, return only the variables needed to index a data array.
+    drop_nan_layers : bool, optional
+        If True, drop the observation wells for which the model layer cannot be
+        determined. These probably lie above or below the model. The default is False.
+        If False, the model layer indexer will contain NaN values for these wells.
+
+    Returns
+    -------
+    obs_ds : xr.Dataset
+        Dataset with the following variables:
+
+        - x: x-coordinate of the cell in which the observation point is located
+        - y: y-coordinate of the cell in which the observation point is located
+        - layer : model layer of each observation point
+
+        Optionally returns the following variables if full_output is True:
+
+        - icell2d: cell id of the observation point
```

Comment:
So in case of a structured grid this is a node number?

Reply:
Correct.